pkg/operation/botanist/health_check.go

```diff
@@ -16,172 +16,422 @@ package botanist
+		if err := health.CheckNode(&object); err != nil {
+			return helper.UpdatedCondition(
+				condition,
+				corev1.ConditionFalse,
+				"NodeUnhealthy",
+				fmt.Sprintf("Node %s is unhealthy: %v", object.Name, err))
+		}
+	}
+
+	return nil
+}
+
+func checkMachineDeployments(condition *gardenv1beta1.Condition, objects []machinev1alpha1.MachineDeployment) *gardenv1beta1.Condition {
+	for _, object := range objects {
+		if err := health.CheckMachineDeployment(&object); err != nil {
+			return helper.UpdatedCondition(
+				condition,
+				corev1.ConditionFalse,
+				"MachineDeploymentUnhealthy",
+				fmt.Sprintf("Machine deployment %s is unhealthy: %v", object.Name, err))
+		}
+	}
+
+	return nil
+}
+
+func checkRequiredDaemonSets(condition *gardenv1beta1.Condition, requiredNames sets.String, objects []appsv1.DaemonSet) *gardenv1beta1.Condition {
+	actualNames := sets.NewString()
+	for _, object := range objects {
+		actualNames.Insert(object.Name)
+	}
+
+	if missingNames := requiredNames.Difference(actualNames); missingNames.Len() != 0 {
+		return helper.UpdatedCondition(
+			condition,
+			corev1.ConditionFalse,
+			"DaemonSetMissing",
+			fmt.Sprintf("Missing required daemon sets: %v", missingNames.List()))
+	}
+	return nil
+}
+
+func checkDaemonSets(condition *gardenv1beta1.Condition, objects []appsv1.DaemonSet) *gardenv1beta1.Condition {
+	for _, object := range objects {
+		if err := health.CheckDaemonSet(&object); err != nil {
+			return helper.UpdatedCondition(
+				condition,
+				corev1.ConditionFalse,
+				"DaemonSetUnhealthy",
+				fmt.Sprintf("Daemon set %s is unhealthy: %v", object.Name, err.Error()))
+		}
+	}
+
+	return nil
+}
+
+func (b *Botanist) checkShootHibernated(condition *gardenv1beta1.Condition) *gardenv1beta1.Condition {
 	if b.Shoot.IsHibernated {
-		return helper.ModifyCondition(condition, corev1.ConditionTrue, "ConditionNotChecked", "Shoot cluster has been hibernated.")
+		return helper.UpdatedCondition(condition, corev1.ConditionTrue, "ConditionNotChecked", "Shoot cluster has been hibernated.")
 	}
+	return nil
+}
 
-	response, err := b.K8sShootClient.Curl("healthz")
+func (b *Botanist) computeRequiredControlPlaneDeployments() sets.String {
+	// This is a hack to quickly do a cloud provider specific check for the required control plane deployments.
+	// As this will anyways change with the Gardener extensibility, for now, this will check for the only
+	// cloud provider where it differs (AWS). Once cloud provider specific code moves out, this will also have to
+	// be refactored / re-aligned.
+	requiredControlPlaneDeployments := sets.NewString(common.RequiredControlPlaneDeployments.UnsortedList()...)
+	if b.Shoot.CloudProvider == gardenv1beta1.CloudProviderAWS {
+		requiredControlPlaneDeployments.Insert(common.AWSLBReadvertiserDeploymentName)
+	}
+	if addons := b.Shoot.Info.Spec.Addons; addons != nil {
+		if clusterAutoscaler := addons.ClusterAutoscaler; clusterAutoscaler != nil && clusterAutoscaler.Enabled {
```

Comment:
This field is deprecated and will be removed in the future.
Better depend on `b.Shoot.WantsAutoscaler`